src/osd/ECBackend.cc

```diff
@@ -590,7 +601,7 @@ void ECBackend::dispatch_recovery_messages(RecoveryMessages &m, int priority)
     m.want_to_read,
     m.recovery_reads,
     OpRequestRef(),
-    false, true);
+    false, true, new RecoveryMessages{});
```

Comment:
Ah, this part is tricky.
IIUC, this part is here solely to make use of RecoveryMessages ability to `dispatch_recovery_messages` on finish.
Can we create a new struct that will serve the same purpose but will allow us to have better differentiation between the two?
RecoveryMessages will be able to reuse that new (acting as on_complete) struct.

Reply:
Moreover, in this case, we use the empty ctor - meaning ec is nullptr and later on we will crash on the call to `ec->dispatch_recovery_messages`

Reply:
Hmm, I'd have thought that defining the single param constructor would have removed the default constructor.  If it's not deleted, let's do so explicitly to avoid doing this by accident.

Reply:
Yeah, I'm with @Matan-B, it's a bit confusing to reuse RecoveryMessages here simply for the 1-line finish() implementation (at least, I don't see the read pipeline using the other bits of RecoveryMessages).

Reply:
Reworked the callback infrastructure around read completion. Now `RecoveryMessage` is a pure struct again.